ibtk/include/ibtk/libmesh_utilities.h

```diff
@@ -169,6 +169,126 @@ copy_and_synch(libMesh::NumericVector<double>& v_in,
     if (close_v_out) v_out.close();
 }
 
+inline void
+batch_vec_copy(const std::vector<libMesh::PetscVector<double>*>& x_vecs,
+               const std::vector<libMesh::PetscVector<double>*>& y_vecs)
+{
+#if defined(NDEBUG)
+    TBOX_ASSERT(x_vecs.size() == y_vecs.size());
+#endif
+    for (unsigned int k = 0; k < x_vecs.size(); ++k)
+    {
+        if (!x_vecs[k] || !y_vecs[k]) continue;
+        int ierr = VecCopy(x_vecs[k]->vec(), y_vecs[k]->vec());
+        IBTK_CHKERRQ(ierr);
+    }
+}
+
+inline void
+batch_vec_copy(const std::vector<std::vector<libMesh::PetscVector<double>*> >& x_vecs,
+               const std::vector<std::vector<libMesh::PetscVector<double>*> >& y_vecs)
+{
+#if defined(NDEBUG)
+    TBOX_ASSERT(x_vecs.size() == y_vecs.size());
+#endif
+    for (unsigned int n = 0; n < x_vecs.size(); ++n)
+    {
+#if defined(NDEBUG)
+        TBOX_ASSERT(x_vecs[n].size() == y_vecs[n].size());
+#endif
+        for (unsigned int k = 0; k < x_vecs[n].size(); ++k)
+        {
+            if (!x_vecs[n][k] || !y_vecs[n][k]) continue;
+            int ierr = VecCopy(x_vecs[n][k]->vec(), y_vecs[n][k]->vec());
+            IBTK_CHKERRQ(ierr);
+        }
+    }
+}
+
+inline void
+batch_vec_assembly(const std::vector<libMesh::PetscVector<double>*>& vecs)
+{
+    for (const auto& v : vecs)
+    {
+        if (!v) continue;
+        int ierr = VecAssemblyBegin(v->vec());
+        IBTK_CHKERRQ(ierr);
+    }
+    for (const auto& v : vecs)
+    {
+        if (!v) continue;
+        int ierr = VecAssemblyEnd(v->vec());
+        IBTK_CHKERRQ(ierr);
+    }
+}
+
+inline void
+batch_vec_assembly(const std::vector<std::vector<libMesh::PetscVector<double>*> >& vecs)
+{
+    for (unsigned int n = 0; n < vecs.size(); ++n)
+    {
+        for (const auto& v : vecs[n])
+        {
+            if (!v) continue;
+            int ierr = VecAssemblyBegin(v->vec());
+            IBTK_CHKERRQ(ierr);
+        }
+    }
+    for (unsigned int n = 0; n < vecs.size(); ++n)
+    {
+        for (const auto& v : vecs[n])
+        {
+            if (!v) continue;
+            int ierr = VecAssemblyEnd(v->vec());
+            IBTK_CHKERRQ(ierr);
+        }
+    }
```

Comment:
same

Reply:
I think we want to do `VecAssemblyBegin` for all `Vec`s before starting `VecAssemblyEnd`.